### ISSUE TITLE
How to add sinkhorn layer and use its gradient in caffe?

### ISSUE THREAD

**Issue by Dark-98** [rfsantacruz/deep-perm-net#1]
Sorry, I didn't find any file about gradient about sinkhorn layer in your code. Can you give me some suggestion? I really need your help. Thanks.

**Comment by Dark-98** [comment]
Sorry, can you help me?

**Comment by rfsantacruz** [comment]
To use this implementation of  sinkhorn layer you just need to put a sequence of row and column normalization in your prototext file like...

layer{
  name: "norm_row1"
  type: "SinkhornNorm"
  bottom: "fc8_s"
  top: "fc8_r1"
  sinkhorn_norm_param{
     is_row_step: true
     num_rows: 9
     num_cols: 9
     lambda: 0.001
  }
}

layer{
  name: "norm_col1"
  type: "SinkhornNorm"
  bottom: "fc8_r1"
  top: "fc8_c1"
  sinkhorn_norm_param{
     is_row_step: false
     num_rows: 9
     num_cols: 9
     lambda: 0.001
  }
}

and the gradient computation is implemented at https://github.com/rfsantacruz/deep-perm-net/blob/master/caffe-perm/src/caffe/layers/sinkhorn_norm_layer.cu. 
However, I suggest you to use new frameworks for deep learning like PyTorch and TensorFlow where the Sinkhorn algorithm is easier to code and use.

**Comment by Dark-98** [comment]
Got it. I will try to use PyTorch. Thanks for your suggestion.
